File: haskell/lojao/src/Product.hs
module Product (
  createProduct,
  Product
) where

  data Product = Product { name :: String,
                            price :: Float,
                            category :: String }

  createProduct :: [Product] -> [IO Product]
  createProduct products = do
    let product = ioCreation
    return ([product] ++ products)
    -- putStrLn "Deseja cadastrar outro produto?"
    -- choice <- getLine
    -- productsNew <- ([product] ++ products)
    -- productsNew
    -- if (createAnother choice) then createProduct newList
    -- else newList
    -- newList

  addProduct :: Product -> [Product] -> [Product]
  addProduct product products = [product] ++ products

  ioCreation :: IO Product
  ioCreation = do
    putStrLn "Qual o nome do produto?"
    name <- getLine
    putStrLn "Digite o preço unitário do produto"
    price <- getLine
    putStrLn "Digite o tipo do produto"
    category <- getLine
    printSuccess name
    return (createNewProduct name (read price :: Float) category)
  
  printSuccess :: String -> IO ()
  printSuccess productName = putStrLn (productName ++ " cadastrado com sucesso!")

  createAnother :: String -> Bool
  createAnother "Sim" = True
  createAnother _ = False

  createNewProduct :: String -> Float -> String -> Product
  createNewProduct name price category = Product { name = name, price = price, category = category }

File: clojure/lojao/src/lojao/product.clj
(ns lojao.product)

(deftype Product [name price category])

(defn new-product [products]
  (println "Digite o nome do produto: ")
  (def name (read-line))
  (println "Digite o preço unitário do produto: ")
  (def price (read-line))
  (println "Digite o tipo do produto:" )
  (def category (read-line))
  (def created-product (Product. name (Double/parseDouble price) category))
  (println "")
  (printf "%s cadastrado com sucesso." name)
  (println "\n")
  (println "Deseja cadastrar outro produto? ")
  (def choice (read-line))
  (def new-products (conj products created-product))
  (if (= choice "Sim") (new-product new-products) new-products))

(defn create-product [products]
  (println "= = = = Cadastro de Produtos = = = =")
  (new-product products))